### PR TITLE
[AI] Unit tests for urlContext APIs

### PR DESCRIPTION
**Pull request by paulb777**
- Add unit tests for the seven mocks added in https://github.com/FirebaseExtended/vertexai-sdk-test-data/pull/54
- Add two additional tests and supporting infra to test empty urlMetadata arrays.

## Summary of Changes

This pull request introduces a robust suite of unit tests for the `urlContext` APIs within both the GoogleAI and VertexAI generative models. The tests cover various response scenarios, ensuring the correct parsing and handling of URL context metadata by the SDK. To support these new tests, the underlying `MockURLProtocol` was enhanced to handle direct `Data` responses, complemented by a new utility function for generating mock JSON HTTP handlers.

### Highlights

* **New Unit Tests for URL Context APIs (GoogleAI)**: Added comprehensive unit tests to `GenerativeModelGoogleAITests.swift` to validate the handling of `urlContext` metadata. These tests cover successful retrieval, mixed validity scenarios (success and error), and cases where URL metadata is empty. A new streaming test for `urlContext` was also introduced.
* **New Unit Tests for URL Context APIs (VertexAI)**: Similar to the GoogleAI tests, new unit tests were added to `GenerativeModelVertexAITests.swift` for `urlContext` APIs. These tests verify successful retrieval, mixed validity, handling of nonexistent retrieved URLs, and empty URL metadata. A streaming test for `urlContext` was also included.
* From #15348 
  *  Added a similar decoding test in GenerateContentResponseTests to verify the same parsing logic when urlMetadata is [] in the Candidate.
  * Added another decoding test in GenerateContentResponseTests to verify that the same behaviour occurs when urlMetadata is omitted (rather than an empty array).
This behaviour is more likely to be observed in practice since [] is typically omitted during JSON serialization in the backend (since it is a default value).

<details>
<summary><b>Changelog</b></summary>

* **FirebaseAI/Tests/Unit/GenerativeModelGoogleAITests.swift**
    * Added `testGenerateContent_success_urlContext` to verify successful URL context retrieval.
    * Added `testGenerateContent_success_urlContext_mixedValidity` to test scenarios with both successful and erroneous URL metadata.
    * Added `testGenerateContent_success_urlContext_emptyURLMetadata` to ensure correct handling when URL metadata is empty.
    * Added `testGenerateContentStream_success_urlContext` to test streaming responses with URL context.
* **FirebaseAI/Tests/Unit/GenerativeModelVertexAITests.swift**
    * Added `testGenerateContent_success_urlContext` to verify successful URL context retrieval.
    * Added `testGenerateContent_success_urlContext_mixedValidity` to test scenarios with both successful and erroneous URL metadata.
    * Added `testGenerateContent_success_urlContext_nonexistentRetrievedURL` to test cases where `retrievedURL` is missing.
    * Added `testGenerateContent_success_urlContext_emptyURLMetadata` to ensure correct handling when URL metadata is empty.
    * Added `testGenerateContentStream_success_urlContext` to test streaming responses with URL context.
* **FirebaseAI/Tests/Unit/MockURLProtocol.swift**
    * Introduced `dataRequestHandler` static variable to allow mocking HTTP responses with raw `Data`.
    * Modified `startLoading` to conditionally use `requestHandler` (for streaming) or `dataRequestHandler` (for static data).
* **FirebaseAI/Tests/Unit/TestUtilities/GenerativeModelTestUtil.swift**
    * Added `httpRequestHandler(json:statusCode:)` utility function to create `URLRequest` handlers that return a specified JSON string as `Data`.
</details>

### CODE REVIEW

File: FirebaseAI/Tests/Unit/GenerativeModelVertexAITests.swift [hunk 1]
```diff
@@ -491,6 +491,70 @@ final class GenerativeModelVertexAITests: XCTestCase {
...
+      withExtension: "json",
+      subdirectory: vertexSubdirectory
+    )
+
+    let response = try await model.generateContent(testPrompt)
+
+    let candidate = try XCTUnwrap(response.candidates.first)
+    let urlContextMetadata = try XCTUnwrap(candidate.urlContextMetadata)
+    let urlMetadata = try XCTUnwrap(urlContextMetadata.urlMetadata.first)
+    XCTAssertEqual(urlMetadata.retrievedURL?.absoluteString, "https://example.com/8")
```
[the remark this answers is not in the document]
Comment (paulb777) [reply FirebaseAI/Tests/Unit/GenerativeModelVertexAITests.swift]: Renamed function to testGenerateContent_success_urlContext_retrievedURLPresentOnErrorStatus cc: @dlarocque @andrewheard

File: FirebaseAI/Tests/Unit/MockURLProtocol.swift
Comment (andrewheard) [inline FirebaseAI/Tests/Unit/MockURLProtocol.swift:1]: To avoid these changes, would you be ok with doing it like this instead? https://github.com/firebase/firebase-ios-sdk/pull/15348

Comment (paulb777) [reply FirebaseAI/Tests/Unit/MockURLProtocol.swift:1]: yes. Merged. Thanks!

**Review by andrewheard — APPROVED** [review]
Thanks, Paul!

**Review by dlarocque — APPROVED** [review]
